ddl/table.go

```diff
@@ -638,19 +638,62 @@ func onRenameTable(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error)
 		return ver, errors.Trace(err)
 	}
 
-	tblInfo, err := getTableInfoAndCancelFaultJob(t, job, oldSchemaID)
-	if err != nil {
-		return ver, errors.Trace(err)
-	}
 	newSchemaID := job.SchemaID
-	err = checkTableNotExists(d, t, newSchemaID, tableName.L)
+	err := checkTableNotExists(d, t, newSchemaID, tableName.L)
 	if err != nil {
 		if infoschema.ErrDatabaseNotExists.Equal(err) || infoschema.ErrTableExists.Equal(err) {
 			job.State = model.JobStateCancelled
 		}
 		return ver, errors.Trace(err)
 	}
 
+	ver, tblInfo, err := checkAndRenameTables(t, job, oldSchemaID, job.SchemaID, &tableName)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	ver, err = updateSchemaVersion(t, job)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
+	return ver, nil
+}
+
+func onRenameTables(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	oldSchemaIDs := []int64{}
+	newSchemaIDs := []int64{}
+	tableNames := []*model.CIStr{}
+	tableIDs := []int64{}
+	if err := job.DecodeArgs(&oldSchemaIDs, &newSchemaIDs, &tableNames, &tableIDs); err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	tblInfo := &model.TableInfo{}
+	var err error
+	for i, oldSchemaID := range oldSchemaIDs {
+		job.TableID = tableIDs[i]
+		ver, tblInfo, err = checkAndRenameTables(t, job, oldSchemaID, newSchemaIDs[i], tableNames[i])
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+	}
```

Comment:
When there's a failure in the middle, all of them need to be rolled back.
Now that a failpoint exists in `checkAndRenameTables`, you can test this case by enabling the failpoint.

Reply:
Should this test case be added in the `TestHalfwayCancelOperations` in `fail_db_test`?

Reply:
Both are OK. I prefer to add it to ddl_test.go.

Reply:
I have added the test case in `fail_db_test`. PTAL